{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d49f13ea-63d6-4128-b606-deeba2daef4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b6c1cb50-74f2-445d-b865-8c22c3b3798b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/work/examples/dfp_workflow/morpheus/morpheus', '/work/examples/dfp_workflow/morpheus', '/opt/conda/envs/morpheus/lib/python38.zip', '/opt/conda/envs/morpheus/lib/python3.8', '/opt/conda/envs/morpheus/lib/python3.8/lib-dynload', '', '/opt/conda/envs/morpheus/lib/python3.8/site-packages']\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "sys.path.insert(0, os.path.abspath(\"./morpheus\"))\n",
    "\n",
    "print(sys.path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "102ce011-3ca3-4f96-a72d-de28fad32003",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/morpheus/lib/python3.8/site-packages/tqdm/auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import logging\n",
    "import os\n",
    "import typing\n",
    "from datetime import datetime\n",
    "from datetime import timedelta\n",
    "\n",
    "import click\n",
    "import mlflow\n",
    "import pandas as pd\n",
    "from dfp.stages.dfp_inference_stage import DFPInferenceStage\n",
    "from dfp.stages.dfp_mlflow_model_writer import DFPMLFlowModelWriterStage\n",
    "from dfp.stages.dfp_postprocessing_stage import DFPPostprocessingStage\n",
    "from dfp.stages.dfp_preprocessing_stage import DFPPreprocessingStage\n",
    "from dfp.stages.dfp_rolling_window_stage import DFPRollingWindowStage\n",
    "from dfp.stages.dfp_s3_batcher_stage import DFPS3BatcherStage\n",
    "from dfp.stages.dfp_s3_to_df import DFPS3ToDataFrameStage\n",
    "from dfp.stages.dfp_split_users_stage import DFPSplitUsersStage\n",
    "from dfp.stages.dfp_training import DFPTraining\n",
    "from dfp.stages.multi_file_source import MultiFileSource\n",
    "from dfp.stages.s3_object_source_stage import S3BucketSourceStage\n",
    "from dfp.stages.s3_object_source_stage import s3_filter_duo\n",
    "from dfp.stages.s3_object_source_stage import s3_object_generator\n",
    "from dfp.stages.write_to_s3_stage import WriteToS3Stage\n",
    "from dfp.utils.column_info import BoolColumn\n",
    "from dfp.utils.column_info import CustomColumn\n",
    "from dfp.utils.column_info import DataFrameInputSchema\n",
    "from dfp.utils.column_info import RenameColumn\n",
    "\n",
    "import cudf\n",
    "\n",
    "from morpheus._lib.file_types import FileTypes\n",
    "from morpheus.config import Config\n",
    "from morpheus.config import ConfigAutoEncoder\n",
    "from morpheus.config import CppConfig\n",
    "from morpheus.messages.message_meta import UserMessageMeta\n",
    "from morpheus.pipeline import LinearPipeline\n",
    "from morpheus.stages.output.write_to_file_stage import WriteToFileStage\n",
    "from morpheus.utils.logger import configure_logging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9ee00703-75c5-46fc-890c-86733da906c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Global options\n",
    "train_users = \"all\"\n",
    "\n",
    "# To include the generic, we must be training all or generic\n",
    "include_generic = train_users == \"all\" or train_users == \"generic\"\n",
    "\n",
    "# To include individual, we must be either training or inferring\n",
    "include_individual = train_users != \"generic\"\n",
    "\n",
    "# None indicates we arent training anything\n",
    "is_training = train_users != \"none\"\n",
    "\n",
    "# Enter any users to skip here\n",
    "skip_users: typing.List[str] = []\n",
    "\n",
    "# Location where cache objects will be saved\n",
    "cache_dir = \"./.cache/dfp\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "01abd537-9162-49dc-8e83-d9465592f1d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enable the Morpheus logger\n",
    "configure_logging(log_level=logging.DEBUG)\n",
    "\n",
    "config = Config()\n",
    "\n",
    "CppConfig.set_should_use_cpp(False)\n",
    "\n",
    "config.num_threads = os.cpu_count()\n",
    "\n",
    "config.ae = ConfigAutoEncoder()\n",
    "\n",
    "config.ae.feature_columns = [\n",
    "    'accessdevicebrowser', 'accessdeviceos', 'device', 'result', 'reason', 'logcount', \"locincrement\"\n",
    "]\n",
    "config.ae.userid_column_name = \"username\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0bc192b5-d4a3-43f5-97c4-38515956206b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def s3_date_extractor_duo(s3_object):\n",
    "    key_object = s3_object.key\n",
    "\n",
    "    # Extract the timestamp from the file name\n",
    "    ts_object = key_object.split('_')[2].split('.json')[0].replace('T', ' ').replace('Z', '')\n",
    "    ts_object = datetime.strptime(ts_object, '%Y-%m-%d %H:%M:%S.%f')\n",
    "\n",
    "    return ts_object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a73a4d53-32b6-4ab8-a5d7-c0104b31c69b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the column names to ensure all data is uniform\n",
    "column_info = [\n",
    "    RenameColumn(name=\"accessdevicebrowser\", dtype=str, input_name=\"access_device.browser\"),\n",
    "    RenameColumn(name=\"accessdeviceos\", dtype=str, input_name=\"access_device.os\"),\n",
    "    RenameColumn(name=\"locationcity\", dtype=str, input_name=\"auth_device.location.city\"),\n",
    "    RenameColumn(name=\"device\", dtype=str, input_name=\"auth_device.name\"),\n",
    "    BoolColumn(name=\"result\",\n",
    "               dtype=bool,\n",
    "               input_name=\"result\",\n",
    "               true_values=[\"success\", \"SUCCESS\"],\n",
    "               false_values=[\"denied\", \"DENIED\", \"FRAUD\"]),\n",
    "    RenameColumn(name=\"reason\", dtype=str, input_name=\"reason\"),\n",
    "    RenameColumn(name=\"username\", dtype=str, input_name=\"user.name\"),\n",
    "    RenameColumn(name=config.ae.timestamp_column_name, dtype=datetime, input_name=config.ae.timestamp_column_name),\n",
    "]\n",
    "\n",
    "input_schema = DataFrameInputSchema(json_columns=[\"access_device\", \"application\", \"auth_device\", \"user\"],\n",
    "                                    column_info=column_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f7a0cb0a-e65a-444a-a06c-a4525d543790",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the final set of columns necessary just before pre-processing\n",
    "def column_logcount(df: cudf.DataFrame):\n",
    "    per_day = df[config.ae.timestamp_column_name].dt.to_period(\"D\")\n",
    "\n",
    "    # Create the per-user, per-day log count\n",
    "    return df.groupby([config.ae.userid_column_name, per_day]).cumcount()\n",
    "\n",
    "def column_locincrement(df: cudf.DataFrame):\n",
    "    per_day = df[config.ae.timestamp_column_name].dt.to_period(\"D\")\n",
    "\n",
    "    # Simple but probably incorrect calculation\n",
    "    return df.groupby([config.ae.userid_column_name, per_day, \"locationcity\"]).ngroup() + 1\n",
    "\n",
    "model_column_info = [\n",
    "    # Input columns\n",
    "    RenameColumn(name=\"accessdevicebrowser\", dtype=str, input_name=\"accessdevicebrowser\"),\n",
    "    RenameColumn(name=\"accessdeviceos\", dtype=str, input_name=\"accessdeviceos\"),\n",
    "    RenameColumn(name=\"device\", dtype=str, input_name=\"device\"),\n",
    "    RenameColumn(name=\"result\", dtype=bool, input_name=\"result\"),\n",
    "    RenameColumn(name=\"reason\", dtype=str, input_name=\"reason\"),\n",
    "    # Derived columns\n",
    "    CustomColumn(name=\"logcount\", dtype=int, process_column_fn=column_logcount),\n",
    "    CustomColumn(name=\"locincrement\", dtype=int, process_column_fn=column_locincrement),\n",
    "    # Extra columns\n",
    "    RenameColumn(name=\"username\", dtype=str, input_name=\"username\"),\n",
    "    RenameColumn(name=config.ae.timestamp_column_name, dtype=datetime, input_name=config.ae.timestamp_column_name),\n",
    "]\n",
    "\n",
    "model_schema = DataFrameInputSchema(column_info=model_column_info, preserve_columns=[\"_batch_id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "825390ad-ce64-4949-b324-33039ffdf264",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "====Registering Pipeline====\u001b[0m\n",
      "====Registering Pipeline Complete!====\u001b[0m\n",
      "====Starting Pipeline====\u001b[0m\n",
      "====Pipeline Started====\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "W20220819 23:30:26.056663  5630 thread.cpp:138] unable to set memory policy - if using docker use: --cap-add=sys_nice to allow membind\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "====Building Pipeline====\u001b[0m\n",
      "Added source: <from-multi-file-0; MultiFileSource(input_schema=DataFrameInputSchema(json_columns=['access_device', 'application', 'auth_device', 'user'], column_info=[RenameColumn(name='accessdevicebrowser', dtype=<class 'str'>, input_name='access_device.browser'), RenameColumn(name='accessdeviceos', dtype=<class 'str'>, input_name='access_device.os'), RenameColumn(name='locationcity', dtype=<class 'str'>, input_name='auth_device.location.city'), RenameColumn(name='device', dtype=<class 'str'>, input_name='auth_device.name'), BoolColumn(name='result', dtype=<class 'bool'>, input_name='result', value_map={'success': True, 'SUCCESS': True, 'denied': False, 'DENIED': False, 'FRAUD': False}), RenameColumn(name='reason', dtype=<class 'str'>, input_name='reason'), RenameColumn(name='username', dtype=<class 'str'>, input_name='user.name'), RenameColumn(name='timestamp', dtype=<class 'datetime.datetime'>, input_name='timestamp')], preserve_columns=None), filenames=['/work/examples/data/dfp/duo/duotest_pt1.json', '/work/examples/data/dfp/duo/duotest_pt2.json', '/work/examples/data/dfp/duo/duotest_pt3.json', '/work/examples/data/dfp/duo/duotest_pt4.json'], file_type=FileTypes.Auto, parser_kwargs={'lines': False, 'orient': 'records'})>\n",
      "  └─> cudf.DataFrame\u001b[0m\n",
      "Added stage: <dfp-split-users-1; DFPSplitUsersStage(include_generic=True, include_individual=True, skip_users=[])>\n",
      "  └─ cudf.DataFrame -> morpheus.UserMessageMeta\u001b[0m\n",
      "Added stage: <dfp-rolling-window-2; DFPRollingWindowStage(min_history=300, min_increment=300, max_history=5d, cache_dir=./.cache/dfp)>\n",
      "  └─ morpheus.UserMessageMeta -> morpheus.UserMessageMeta\u001b[0m\n",
      "Added stage: <dfp-preproc-3; DFPPreprocessingStage(input_schema=DataFrameInputSchema(json_columns=[], column_info=[RenameColumn(name='accessdevicebrowser', dtype=<class 'str'>, input_name='accessdevicebrowser'), RenameColumn(name='accessdeviceos', dtype=<class 'str'>, input_name='accessdeviceos'), RenameColumn(name='device', dtype=<class 'str'>, input_name='device'), RenameColumn(name='result', dtype=<class 'bool'>, input_name='result'), RenameColumn(name='reason', dtype=<class 'str'>, input_name='reason'), CustomColumn(name='logcount', dtype=<class 'int'>, process_column_fn=<function column_logcount at 0x7f8ea87003a0>), CustomColumn(name='locincrement', dtype=<class 'int'>, process_column_fn=<function column_locincrement at 0x7f8eac401310>), RenameColumn(name='username', dtype=<class 'str'>, input_name='username'), RenameColumn(name='timestamp', dtype=<class 'datetime.datetime'>, input_name='timestamp')], preserve_columns=re.compile('(_batch_id)')), return_format=data, only_last_batch=False)>\n",
      "  └─ morpheus.UserMessageMeta -> morpheus.UserMessageMeta\u001b[0m\n",
      "Added stage: <dfp-training-4; DFPTraining()>\n",
      "  └─ morpheus.UserMessageMeta -> morpheus.MultiAEMessage\u001b[0m\n",
      "Added stage: <dfp-mlflow-model-writer-5; DFPMLFlowModelWriterStage()>\n",
      "  └─ morpheus.MultiAEMessage -> morpheus.UserMessageMeta\u001b[0m\n",
      "====Building Pipeline Complete!====\u001b[0m\n",
      "\u001b[2mStarting! Time: 1660951826.0659451\u001b[0m\n",
      "Sending 4000 rows\n",
      "Got message:     accessdevicebrowser accessdeviceos locationcity           device  result  \\\n",
      "0                  null           null          NaN  +91 79876 42470    True   \n",
      "1                                               NaN                     True   \n",
      "2                  null           null          NaN   +852 9147 4309    True   \n",
      "3                                               NaN                     True   \n",
      "4                                               NaN                     True   \n",
      "..                  ...            ...          ...              ...     ...   \n",
      "995                                             NaN                     True   \n",
      "996                                             NaN                     True   \n",
      "997                                             NaN                     True   \n",
      "998                                             NaN                     True   \n",
      "999                                             NaN                     True   \n",
      "\n",
      "            reason        username           timestamp  \n",
      "0    User approved          badguy 2021-09-14 00:52:14  \n",
      "1                           badguy 2021-09-14 07:04:46  \n",
      "2    User approved  maliciousactor 2021-09-14 16:58:15  \n",
      "3                   maliciousactor 2021-09-15 07:12:29  \n",
      "4                   maliciousactor 2021-09-15 07:17:55  \n",
      "..             ...             ...                 ...  \n",
      "995                          userh 2022-01-24 15:22:10  \n",
      "996                          userb 2022-01-24 15:22:10  \n",
      "997                          userc 2022-01-24 15:23:00  \n",
      "998                          userg 2022-01-24 15:23:25  \n",
      "999                 maliciousactor 2022-01-24 15:23:41  \n",
      "\n",
      "[4000 rows x 8 columns]\n",
      "\u001b[2mBatch split users complete. Input: 4000 rows from 2021-09-14 00:52:14 to 2022-01-24 15:23:41. Output: 11 users, rows/user min: 176, max: 4000, avg: 727.27. Duration: 3.51 ms\u001b[0m\n",
      "\u001b[2mRolling window complete for generic_user in 14.32 ms. Input: 4000 rows from 2021-09-14 00:52:14 to 2022-01-24 15:23:41. Output: 4000 rows from 2021-09-14 00:52:14 to 2022-01-24 15:23:41\u001b[0m\n",
      "\u001b[2mRolling window complete for badguy in 9.74 ms. Input: 490 rows from 2021-09-14 00:52:14 to 2022-01-24 09:30:26. Output: 490 rows from 2021-09-14 00:52:14 to 2022-01-24 09:30:26\u001b[0m\n",
      "\u001b[2mRolling window complete for maliciousactor in 9.46 ms. Input: 700 rows from 2021-09-14 16:58:15 to 2022-01-24 15:23:41. Output: 700 rows from 2021-09-14 16:58:15 to 2022-01-24 15:23:41\u001b[0m\n",
      "\u001b[2mRolling window complete for userc in 9.60 ms. Input: 500 rows from 2021-10-06 20:29:50 to 2022-01-24 15:23:00. Output: 500 rows from 2021-10-06 20:29:50 to 2022-01-24 15:23:00\u001b[0m\n",
      "\u001b[2mRolling window complete for userd in 12.76 ms. Input: 537 rows from 2021-10-07 02:25:27 to 2022-01-24 12:26:34. Output: 537 rows from 2021-10-07 02:25:27 to 2022-01-24 12:26:34\u001b[0m\n",
      "\u001b[2mRolling window complete for userb in 10.21 ms. Input: 457 rows from 2021-10-07 04:02:07 to 2022-01-24 15:22:10. Output: 457 rows from 2021-10-07 04:02:07 to 2022-01-24 15:22:10\u001b[0m\n",
      "\u001b[2mRolling window complete for userh in 7.64 ms. Input: 346 rows from 2021-10-07 10:54:07 to 2022-01-24 15:22:10. Output: 346 rows from 2021-10-07 10:54:07 to 2022-01-24 15:22:10\u001b[0m\n",
      "\u001b[2mPreprocessed 4000 data for logs in 2021-09-14 00:52:14 to 2022-01-24 15:23:41 in 68.42231750488281 ms\u001b[0m\n",
      "\u001b[2mRolling window complete for userf in 10.65 ms. Input: 304 rows from 2021-10-07 10:55:23 to 2022-01-24 14:08:00. Output: 304 rows from 2021-10-07 10:55:23 to 2022-01-24 14:08:00\u001b[0m\n",
      "\u001b[2mPreprocessed 490 data for logs in 2021-09-14 00:52:14 to 2022-01-24 09:30:26 in 31.4176082611084 ms\u001b[0m\n",
      "\u001b[2mPreprocessed 700 data for logs in 2021-09-14 16:58:15 to 2022-01-24 15:23:41 in 20.436763763427734 ms\u001b[0m\n",
      "\u001b[2mTraining AE model for user: 'generic_user'...\u001b[0m\n",
      "\u001b[2mPreprocessed 500 data for logs in 2021-10-06 20:29:50 to 2022-01-24 15:23:00 in 21.993637084960938 ms\u001b[0m\n",
      "\u001b[2mPreprocessed 537 data for logs in 2021-10-07 02:25:27 to 2022-01-24 12:26:34 in 14.673233032226562 ms\u001b[0m\n",
      "\u001b[2mPreprocessed 457 data for logs in 2021-10-07 04:02:07 to 2022-01-24 15:22:10 in 4064.789295196533 ms\u001b[0m\n",
      "\u001b[2mPreprocessed 346 data for logs in 2021-10-07 10:54:07 to 2022-01-24 15:22:10 in 31.79931640625 ms\u001b[0m\n",
      "\u001b[2mPreprocessed 304 data for logs in 2021-10-07 10:55:23 to 2022-01-24 14:08:00 in 39.21818733215332 ms\u001b[0m\n",
      "\u001b[2mTraining AE model for user: 'generic_user'... Complete.\u001b[0m\n",
      "\u001b[2mTraining AE model for user: 'badguy'...\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[31mError trying to upload ML Flow model\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/registry.py\", line 76, in get_store_builder\n",
      "    store_builder = self._registry[scheme]\n",
      "KeyError: 'file'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 95, in _get_registry_client\n",
      "    registry_client = ModelRegistryClient(self._registry_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/client.py\", line 36, in __init__\n",
      "    self.store  # pylint: disable=pointless-statement\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/client.py\", line 40, in store\n",
      "    return utils._get_store(self.registry_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/utils.py\", line 169, in _get_store\n",
      "    return _get_store_registry().get_store(store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/registry.py\", line 37, in get_store\n",
      "    return self._get_store_with_resolved_uri(resolved_store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/registry.py\", line 46, in _get_store_with_resolved_uri\n",
      "    builder = self.get_store_builder(resolved_store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/registry.py\", line 78, in get_store_builder\n",
      "    raise UnsupportedModelRegistryStoreURIException(\n",
      "mlflow.tracking.registry.UnsupportedModelRegistryStoreURIException:  Model registry functionality is unavailable; got unsupported URI 'file:///work/examples/dfp_workflow/morpheus/mlruns' for model registry data storage. Supported URI schemes are: ['databricks', 'http', 'https', 'postgresql', 'mysql', 'sqlite', 'mssql']. See https://www.mlflow.org/docs/latest/tracking.html#storage for how to run an MLflow server against one of the supported backend storage locations.\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/work/examples/dfp_workflow/morpheus/dfp/stages/dfp_mlflow_model_writer.py\", line 142, in on_data\n",
      "    raise e\n",
      "  File \"/work/examples/dfp_workflow/morpheus/dfp/stages/dfp_mlflow_model_writer.py\", line 136, in on_data\n",
      "    create_model_response = client.create_registered_model(reg_model_name)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 1666, in create_registered_model\n",
      "    return self._get_registry_client().create_registered_model(name, tags, description)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 102, in _get_registry_client\n",
      "    raise MlflowException(\n",
      "mlflow.exceptions.MlflowException: Model Registry features are not supported by the store with URI: 'file:///work/examples/dfp_workflow/morpheus/mlruns'. Stores with the following URI schemes are supported: ['databricks', 'http', 'https', 'postgresql', 'mysql', 'sqlite', 'mssql'].\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2mTraining AE model for user: 'badguy'... Complete.\u001b[0m\n",
      "\u001b[2mTraining AE model for user: 'maliciousactor'...\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[31mError trying to upload ML Flow model\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/registry.py\", line 76, in get_store_builder\n",
      "    store_builder = self._registry[scheme]\n",
      "KeyError: 'file'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 95, in _get_registry_client\n",
      "    registry_client = ModelRegistryClient(self._registry_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/client.py\", line 36, in __init__\n",
      "    self.store  # pylint: disable=pointless-statement\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/client.py\", line 40, in store\n",
      "    return utils._get_store(self.registry_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/utils.py\", line 169, in _get_store\n",
      "    return _get_store_registry().get_store(store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/registry.py\", line 37, in get_store\n",
      "    return self._get_store_with_resolved_uri(resolved_store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/registry.py\", line 46, in _get_store_with_resolved_uri\n",
      "    builder = self.get_store_builder(resolved_store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/registry.py\", line 78, in get_store_builder\n",
      "    raise UnsupportedModelRegistryStoreURIException(\n",
      "mlflow.tracking.registry.UnsupportedModelRegistryStoreURIException:  Model registry functionality is unavailable; got unsupported URI 'file:///work/examples/dfp_workflow/morpheus/mlruns' for model registry data storage. Supported URI schemes are: ['databricks', 'http', 'https', 'postgresql', 'mysql', 'sqlite', 'mssql']. See https://www.mlflow.org/docs/latest/tracking.html#storage for how to run an MLflow server against one of the supported backend storage locations.\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/work/examples/dfp_workflow/morpheus/dfp/stages/dfp_mlflow_model_writer.py\", line 142, in on_data\n",
      "    raise e\n",
      "  File \"/work/examples/dfp_workflow/morpheus/dfp/stages/dfp_mlflow_model_writer.py\", line 136, in on_data\n",
      "    create_model_response = client.create_registered_model(reg_model_name)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 1666, in create_registered_model\n",
      "    return self._get_registry_client().create_registered_model(name, tags, description)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 102, in _get_registry_client\n",
      "    raise MlflowException(\n",
      "mlflow.exceptions.MlflowException: Model Registry features are not supported by the store with URI: 'file:///work/examples/dfp_workflow/morpheus/mlruns'. Stores with the following URI schemes are supported: ['databricks', 'http', 'https', 'postgresql', 'mysql', 'sqlite', 'mssql'].\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2mTraining AE model for user: 'maliciousactor'... Complete.\u001b[0m\n",
      "\u001b[2mTraining AE model for user: 'userc'...\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[31mError trying to upload ML Flow model\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/registry.py\", line 76, in get_store_builder\n",
      "    store_builder = self._registry[scheme]\n",
      "KeyError: 'file'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 95, in _get_registry_client\n",
      "    registry_client = ModelRegistryClient(self._registry_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/client.py\", line 36, in __init__\n",
      "    self.store  # pylint: disable=pointless-statement\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/client.py\", line 40, in store\n",
      "    return utils._get_store(self.registry_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/utils.py\", line 169, in _get_store\n",
      "    return _get_store_registry().get_store(store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/registry.py\", line 37, in get_store\n",
      "    return self._get_store_with_resolved_uri(resolved_store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/registry.py\", line 46, in _get_store_with_resolved_uri\n",
      "    builder = self.get_store_builder(resolved_store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/registry.py\", line 78, in get_store_builder\n",
      "    raise UnsupportedModelRegistryStoreURIException(\n",
      "mlflow.tracking.registry.UnsupportedModelRegistryStoreURIException:  Model registry functionality is unavailable; got unsupported URI 'file:///work/examples/dfp_workflow/morpheus/mlruns' for model registry data storage. Supported URI schemes are: ['databricks', 'http', 'https', 'postgresql', 'mysql', 'sqlite', 'mssql']. See https://www.mlflow.org/docs/latest/tracking.html#storage for how to run an MLflow server against one of the supported backend storage locations.\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/work/examples/dfp_workflow/morpheus/dfp/stages/dfp_mlflow_model_writer.py\", line 142, in on_data\n",
      "    raise e\n",
      "  File \"/work/examples/dfp_workflow/morpheus/dfp/stages/dfp_mlflow_model_writer.py\", line 136, in on_data\n",
      "    create_model_response = client.create_registered_model(reg_model_name)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 1666, in create_registered_model\n",
      "    return self._get_registry_client().create_registered_model(name, tags, description)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 102, in _get_registry_client\n",
      "    raise MlflowException(\n",
      "mlflow.exceptions.MlflowException: Model Registry features are not supported by the store with URI: 'file:///work/examples/dfp_workflow/morpheus/mlruns'. Stores with the following URI schemes are supported: ['databricks', 'http', 'https', 'postgresql', 'mysql', 'sqlite', 'mssql'].\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2mTraining AE model for user: 'userc'... Complete.\u001b[0m\n",
      "\u001b[2mTraining AE model for user: 'userd'...\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[31mError trying to upload ML Flow model\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/registry.py\", line 76, in get_store_builder\n",
      "    store_builder = self._registry[scheme]\n",
      "KeyError: 'file'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 95, in _get_registry_client\n",
      "    registry_client = ModelRegistryClient(self._registry_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/client.py\", line 36, in __init__\n",
      "    self.store  # pylint: disable=pointless-statement\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/client.py\", line 40, in store\n",
      "    return utils._get_store(self.registry_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/utils.py\", line 169, in _get_store\n",
      "    return _get_store_registry().get_store(store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/registry.py\", line 37, in get_store\n",
      "    return self._get_store_with_resolved_uri(resolved_store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/registry.py\", line 46, in _get_store_with_resolved_uri\n",
      "    builder = self.get_store_builder(resolved_store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/registry.py\", line 78, in get_store_builder\n",
      "    raise UnsupportedModelRegistryStoreURIException(\n",
      "mlflow.tracking.registry.UnsupportedModelRegistryStoreURIException:  Model registry functionality is unavailable; got unsupported URI 'file:///work/examples/dfp_workflow/morpheus/mlruns' for model registry data storage. Supported URI schemes are: ['databricks', 'http', 'https', 'postgresql', 'mysql', 'sqlite', 'mssql']. See https://www.mlflow.org/docs/latest/tracking.html#storage for how to run an MLflow server against one of the supported backend storage locations.\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/work/examples/dfp_workflow/morpheus/dfp/stages/dfp_mlflow_model_writer.py\", line 142, in on_data\n",
      "    raise e\n",
      "  File \"/work/examples/dfp_workflow/morpheus/dfp/stages/dfp_mlflow_model_writer.py\", line 136, in on_data\n",
      "    create_model_response = client.create_registered_model(reg_model_name)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 1666, in create_registered_model\n",
      "    return self._get_registry_client().create_registered_model(name, tags, description)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 102, in _get_registry_client\n",
      "    raise MlflowException(\n",
      "mlflow.exceptions.MlflowException: Model Registry features are not supported by the store with URI: 'file:///work/examples/dfp_workflow/morpheus/mlruns'. Stores with the following URI schemes are supported: ['databricks', 'http', 'https', 'postgresql', 'mysql', 'sqlite', 'mssql'].\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2mTraining AE model for user: 'userd'... Complete.\u001b[0m\n",
      "\u001b[2mTraining AE model for user: 'userb'...\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[31mError trying to upload ML Flow model\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/registry.py\", line 76, in get_store_builder\n",
      "    store_builder = self._registry[scheme]\n",
      "KeyError: 'file'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 95, in _get_registry_client\n",
      "    registry_client = ModelRegistryClient(self._registry_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/client.py\", line 36, in __init__\n",
      "    self.store  # pylint: disable=pointless-statement\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/client.py\", line 40, in store\n",
      "    return utils._get_store(self.registry_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/utils.py\", line 169, in _get_store\n",
      "    return _get_store_registry().get_store(store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/registry.py\", line 37, in get_store\n",
      "    return self._get_store_with_resolved_uri(resolved_store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/registry.py\", line 46, in _get_store_with_resolved_uri\n",
      "    builder = self.get_store_builder(resolved_store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/registry.py\", line 78, in get_store_builder\n",
      "    raise UnsupportedModelRegistryStoreURIException(\n",
      "mlflow.tracking.registry.UnsupportedModelRegistryStoreURIException:  Model registry functionality is unavailable; got unsupported URI 'file:///work/examples/dfp_workflow/morpheus/mlruns' for model registry data storage. Supported URI schemes are: ['databricks', 'http', 'https', 'postgresql', 'mysql', 'sqlite', 'mssql']. See https://www.mlflow.org/docs/latest/tracking.html#storage for how to run an MLflow server against one of the supported backend storage locations.\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/work/examples/dfp_workflow/morpheus/dfp/stages/dfp_mlflow_model_writer.py\", line 142, in on_data\n",
      "    raise e\n",
      "  File \"/work/examples/dfp_workflow/morpheus/dfp/stages/dfp_mlflow_model_writer.py\", line 136, in on_data\n",
      "    create_model_response = client.create_registered_model(reg_model_name)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 1666, in create_registered_model\n",
      "    return self._get_registry_client().create_registered_model(name, tags, description)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 102, in _get_registry_client\n",
      "    raise MlflowException(\n",
      "mlflow.exceptions.MlflowException: Model Registry features are not supported by the store with URI: 'file:///work/examples/dfp_workflow/morpheus/mlruns'. Stores with the following URI schemes are supported: ['databricks', 'http', 'https', 'postgresql', 'mysql', 'sqlite', 'mssql'].\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2mTraining AE model for user: 'userb'... Complete.\u001b[0m\n",
      "\u001b[2mTraining AE model for user: 'userh'...\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[31mError trying to upload ML Flow model\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/registry.py\", line 76, in get_store_builder\n",
      "    store_builder = self._registry[scheme]\n",
      "KeyError: 'file'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 95, in _get_registry_client\n",
      "    registry_client = ModelRegistryClient(self._registry_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/client.py\", line 36, in __init__\n",
      "    self.store  # pylint: disable=pointless-statement\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/client.py\", line 40, in store\n",
      "    return utils._get_store(self.registry_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/utils.py\", line 169, in _get_store\n",
      "    return _get_store_registry().get_store(store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/registry.py\", line 37, in get_store\n",
      "    return self._get_store_with_resolved_uri(resolved_store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/registry.py\", line 46, in _get_store_with_resolved_uri\n",
      "    builder = self.get_store_builder(resolved_store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/registry.py\", line 78, in get_store_builder\n",
      "    raise UnsupportedModelRegistryStoreURIException(\n",
      "mlflow.tracking.registry.UnsupportedModelRegistryStoreURIException:  Model registry functionality is unavailable; got unsupported URI 'file:///work/examples/dfp_workflow/morpheus/mlruns' for model registry data storage. Supported URI schemes are: ['databricks', 'http', 'https', 'postgresql', 'mysql', 'sqlite', 'mssql']. See https://www.mlflow.org/docs/latest/tracking.html#storage for how to run an MLflow server against one of the supported backend storage locations.\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/work/examples/dfp_workflow/morpheus/dfp/stages/dfp_mlflow_model_writer.py\", line 142, in on_data\n",
      "    raise e\n",
      "  File \"/work/examples/dfp_workflow/morpheus/dfp/stages/dfp_mlflow_model_writer.py\", line 136, in on_data\n",
      "    create_model_response = client.create_registered_model(reg_model_name)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 1666, in create_registered_model\n",
      "    return self._get_registry_client().create_registered_model(name, tags, description)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 102, in _get_registry_client\n",
      "    raise MlflowException(\n",
      "mlflow.exceptions.MlflowException: Model Registry features are not supported by the store with URI: 'file:///work/examples/dfp_workflow/morpheus/mlruns'. Stores with the following URI schemes are supported: ['databricks', 'http', 'https', 'postgresql', 'mysql', 'sqlite', 'mssql'].\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2mTraining AE model for user: 'userh'... Complete.\u001b[0m\n",
      "\u001b[2mTraining AE model for user: 'userf'...\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[31mError trying to upload ML Flow model\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/registry.py\", line 76, in get_store_builder\n",
      "    store_builder = self._registry[scheme]\n",
      "KeyError: 'file'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 95, in _get_registry_client\n",
      "    registry_client = ModelRegistryClient(self._registry_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/client.py\", line 36, in __init__\n",
      "    self.store  # pylint: disable=pointless-statement\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/client.py\", line 40, in store\n",
      "    return utils._get_store(self.registry_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/utils.py\", line 169, in _get_store\n",
      "    return _get_store_registry().get_store(store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/registry.py\", line 37, in get_store\n",
      "    return self._get_store_with_resolved_uri(resolved_store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/registry.py\", line 46, in _get_store_with_resolved_uri\n",
      "    builder = self.get_store_builder(resolved_store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/registry.py\", line 78, in get_store_builder\n",
      "    raise UnsupportedModelRegistryStoreURIException(\n",
      "mlflow.tracking.registry.UnsupportedModelRegistryStoreURIException:  Model registry functionality is unavailable; got unsupported URI 'file:///work/examples/dfp_workflow/morpheus/mlruns' for model registry data storage. Supported URI schemes are: ['databricks', 'http', 'https', 'postgresql', 'mysql', 'sqlite', 'mssql']. See https://www.mlflow.org/docs/latest/tracking.html#storage for how to run an MLflow server against one of the supported backend storage locations.\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/work/examples/dfp_workflow/morpheus/dfp/stages/dfp_mlflow_model_writer.py\", line 142, in on_data\n",
      "    raise e\n",
      "  File \"/work/examples/dfp_workflow/morpheus/dfp/stages/dfp_mlflow_model_writer.py\", line 136, in on_data\n",
      "    create_model_response = client.create_registered_model(reg_model_name)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 1666, in create_registered_model\n",
      "    return self._get_registry_client().create_registered_model(name, tags, description)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 102, in _get_registry_client\n",
      "    raise MlflowException(\n",
      "mlflow.exceptions.MlflowException: Model Registry features are not supported by the store with URI: 'file:///work/examples/dfp_workflow/morpheus/mlruns'. Stores with the following URI schemes are supported: ['databricks', 'http', 'https', 'postgresql', 'mysql', 'sqlite', 'mssql'].\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2mTraining AE model for user: 'userf'... Complete.\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[31mError trying to upload ML Flow model\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/registry.py\", line 76, in get_store_builder\n",
      "    store_builder = self._registry[scheme]\n",
      "KeyError: 'file'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 95, in _get_registry_client\n",
      "    registry_client = ModelRegistryClient(self._registry_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/client.py\", line 36, in __init__\n",
      "    self.store  # pylint: disable=pointless-statement\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/client.py\", line 40, in store\n",
      "    return utils._get_store(self.registry_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/utils.py\", line 169, in _get_store\n",
      "    return _get_store_registry().get_store(store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/registry.py\", line 37, in get_store\n",
      "    return self._get_store_with_resolved_uri(resolved_store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/_model_registry/registry.py\", line 46, in _get_store_with_resolved_uri\n",
      "    builder = self.get_store_builder(resolved_store_uri)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/registry.py\", line 78, in get_store_builder\n",
      "    raise UnsupportedModelRegistryStoreURIException(\n",
      "mlflow.tracking.registry.UnsupportedModelRegistryStoreURIException:  Model registry functionality is unavailable; got unsupported URI 'file:///work/examples/dfp_workflow/morpheus/mlruns' for model registry data storage. Supported URI schemes are: ['databricks', 'http', 'https', 'postgresql', 'mysql', 'sqlite', 'mssql']. See https://www.mlflow.org/docs/latest/tracking.html#storage for how to run an MLflow server against one of the supported backend storage locations.\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/work/examples/dfp_workflow/morpheus/dfp/stages/dfp_mlflow_model_writer.py\", line 142, in on_data\n",
      "    raise e\n",
      "  File \"/work/examples/dfp_workflow/morpheus/dfp/stages/dfp_mlflow_model_writer.py\", line 136, in on_data\n",
      "    create_model_response = client.create_registered_model(reg_model_name)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 1666, in create_registered_model\n",
      "    return self._get_registry_client().create_registered_model(name, tags, description)\n",
      "  File \"/opt/conda/envs/morpheus/lib/python3.8/site-packages/mlflow/tracking/client.py\", line 102, in _get_registry_client\n",
      "    raise MlflowException(\n",
      "mlflow.exceptions.MlflowException: Model Registry features are not supported by the store with URI: 'file:///work/examples/dfp_workflow/morpheus/mlruns'. Stores with the following URI schemes are supported: ['databricks', 'http', 'https', 'postgresql', 'mysql', 'sqlite', 'mssql'].\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "====Pipeline Complete====\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "from dfp.stages.dfp_inference_stage import DFPInferenceStage\n",
    "from dfp.stages.dfp_mlflow_model_writer import DFPMLFlowModelWriterStage\n",
    "from dfp.stages.dfp_postprocessing_stage import DFPPostprocessingStage\n",
    "from dfp.stages.dfp_preprocessing_stage import DFPPreprocessingStage\n",
    "from dfp.stages.dfp_rolling_window_stage import DFPRollingWindowStage\n",
    "from dfp.stages.dfp_s3_batcher_stage import DFPS3BatcherStage\n",
    "from dfp.stages.dfp_s3_to_df import DFPS3ToDataFrameStage\n",
    "from dfp.stages.dfp_split_users_stage import DFPSplitUsersStage\n",
    "from dfp.stages.dfp_training import DFPTraining\n",
    "from dfp.stages.multi_file_source import MultiFileSource\n",
    "from dfp.stages.s3_object_source_stage import S3BucketSourceStage\n",
    "from dfp.stages.s3_object_source_stage import s3_filter_duo\n",
    "from dfp.stages.s3_object_source_stage import s3_object_generator\n",
    "from dfp.stages.write_to_s3_stage import WriteToS3Stage\n",
    "\n",
    "# Create a linear pipeline object\n",
    "pipeline = LinearPipeline(config)\n",
    "\n",
    "# Source stage uses \n",
    "pipeline.set_source(\n",
    "    MultiFileSource(config,\n",
    "                    input_schema=input_schema,\n",
    "                    filenames=[\"/work/examples/data/dfp/duo/duotest_pt1.json\", \"/work/examples/data/dfp/duo/duotest_pt2.json\", \"/work/examples/data/dfp/duo/duotest_pt3.json\", \"/work/examples/data/dfp/duo/duotest_pt4.json\"],\n",
    "                    parser_kwargs={\n",
    "                        \"lines\": False, \"orient\": \"records\"\n",
    "                    }))\n",
    "\n",
    "# This will split users or just use one single user\n",
    "pipeline.add_stage(\n",
    "    DFPSplitUsersStage(config,\n",
    "                       include_generic=include_generic,\n",
    "                       include_individual=include_individual,\n",
    "                       skip_users=skip_users))\n",
    "\n",
    "# Next, have a stage that will create rolling windows\n",
    "pipeline.add_stage(\n",
    "    DFPRollingWindowStage(\n",
    "        config,\n",
    "        min_history=300 if is_training else 1,\n",
    "        min_increment=300 if is_training else 1,\n",
    "        # For inference, we only ever want 1 day max\n",
    "        max_history=\"5d\" if is_training else \"1d\",\n",
    "        cache_dir=cache_dir))\n",
    "\n",
    "# Specify the final set of columns necessary just before pre-processing\n",
    "model_column_info = [\n",
    "    # Input columns\n",
    "    RenameColumn(name=\"accessdevicebrowser\", dtype=str, input_name=\"accessdevicebrowser\"),\n",
    "    RenameColumn(name=\"accessdeviceos\", dtype=str, input_name=\"accessdeviceos\"),\n",
    "    RenameColumn(name=\"device\", dtype=str, input_name=\"device\"),\n",
    "    RenameColumn(name=\"result\", dtype=bool, input_name=\"result\"),\n",
    "    RenameColumn(name=\"reason\", dtype=str, input_name=\"reason\"),\n",
    "    # Derived columns\n",
    "    CustomColumn(name=\"logcount\", dtype=int, process_column_fn=column_logcount),\n",
    "    CustomColumn(name=\"locincrement\", dtype=int, process_column_fn=column_locincrement),\n",
    "    # Extra columns\n",
    "    RenameColumn(name=\"username\", dtype=str, input_name=\"username\"),\n",
    "    RenameColumn(name=config.ae.timestamp_column_name, dtype=datetime, input_name=config.ae.timestamp_column_name),\n",
    "]\n",
    "\n",
    "model_schema = DataFrameInputSchema(column_info=model_column_info, preserve_columns=[\"_batch_id\"])\n",
    "\n",
    "# Output is UserMessageMeta -- Cached frame set\n",
    "pipeline.add_stage(DFPPreprocessingStage(config, input_schema=model_schema, only_last_batch=not is_training))\n",
    "\n",
    "if (is_training):\n",
    "\n",
    "    pipeline.add_stage(DFPTraining(config))\n",
    "\n",
    "    pipeline.add_stage(DFPMLFlowModelWriterStage(config))\n",
    "else:\n",
    "    pipeline.add_stage(DFPInferenceStage(config))\n",
    "\n",
    "    pipeline.add_stage(DFPPostprocessingStage(config))\n",
    "\n",
    "    pipeline.add_stage(WriteToFileStage(config, filename=\"dfp_detections.csv\", overwrite=True))\n",
    "\n",
    "# Run the pipeline\n",
    "await pipeline._do_run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52c848c1-d56d-42a0-83ad-19d6fe9731a6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6430831c-03c0-4cfb-8688-8777870ca722",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:morpheus] *",
   "language": "python",
   "name": "conda-env-morpheus-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "vscode": {
   "interpreter": {
    "hash": "e26783b24f020aa0bcaa00e6ba122db5d0e3da2d892d80be664969895e06a7e1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
